### PR TITLE
Introduce table OID

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB

This PR introduces OIDs to tables which allows identifying relations by `OID` instead of s`chema + name`. This is a pre-requisite to make sure concurrent operations like table swaps or renames don't break queries that are already running.

Key changes:
- Added an int `tableOID` field to `RelationMetadata`, `DocTableInfo` to uniquely identify tables.
- Added a `Metadata.currentMaxTableOid`, `Metadata.Builder.currentMaxTableOid`, and `Metadata.Builder.tableOidSupplier` to transfer the current max table oid when building a new `Metadata`.

`Metadata` exposes `Metadata.Builder` with two factory methods. Updated them to make sure the `OidSupplier` is in sync:
- `builder(Metadata metadata)`: This is for normal updates. It just carries over the current supplier to the new `Metadata`.
- `builder()`: This has more complex uses:
  - bootstrapping
  - loading from disk
  - logical replication communication
  - restoring snapshot
  

To make sure that `builder()` keeps the consistency of the OIDs, it has been updated to accept `currentMaxTableOid` parameter, then added an assert in `Metadata.Builder.build()`, it checks:
- all assigned tableOIDs are unique.
- All assigned tableOIDs are <= tableOidSupplier

**Exception**: There is an exception for restoring old Metadata. The restore process creates Metadata twice (once to restore, once to upgrade). The restored one might fail the assert until the upgrade is finished. This is only for version >= 6 where RelationMetadata was first introduced.

Follow ups:
- update relation lookups to actually use these OIDs instead of names.
- add the actual logic to handle concurrent table swaps and renames.
- https://github.com/crate/crate/pull/18954#discussion_r2742325630 - could also be a separate FR.
## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by jeeminso** [comment]
I thought about testing swap/rename on tables created on the latest master, after 6.0, and before 6.0; but it seems more suited as a `crate-qa` test.

**Comment by jeeminso** [comment]
Could you take a look @mfussenegger @seut ? I still need to clean up a bit more and add tests but would like to get some feedback before I proceed further.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -507,9 +517,14 @@ private static class MetadataDiff implements Diff<Metadata> {
             } else {
                 columnOID = COLUMN_OID_UNASSIGNED;
             }
+            if (in.getVersion().after(Version.V_6_2_0)) {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]:
```suggestion
            if (in.getVersion().after(Version.V_6_3_0)) {
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 2]
```diff
@@ -533,9 +548,12 @@ public void writeTo(StreamOutput out) throws IOException {
             if (hasGlobalColumnOID(out.getVersion())) {
                 out.writeLong(columnOID);
             }
+            if (out.getVersion().after(Version.V_6_2_0)) {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]:
```suggestion
            if (out.getVersion().after(Version.V_6_3_0)) {
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 3]
```diff
@@ -563,11 +582,16 @@ public Metadata apply(Metadata part) {
     }
 
     public static Metadata readFrom(StreamInput in) throws IOException {
-        Builder builder = new Builder();
+        Builder builder = Metadata.builder();
         builder.version = in.readLong();
         if (hasGlobalColumnOID(in.getVersion())) {
             builder.columnOID(in.readLong());
         }
+        if (in.getVersion().after(Version.V_6_2_0)) {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]:
```suggestion
        if (in.getVersion().after(Version.V_6_3_0)) {
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 4]
```diff
@@ -609,6 +633,9 @@ public void writeTo(StreamOutput out) throws IOException {
         if (hasGlobalColumnOID(out.getVersion())) {
             out.writeLong(columnOID);
         }
+        if (out.getVersion().after(Version.V_6_2_0)) {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]:
```suggestion
        if (out.getVersion().after(Version.V_6_3_0)) {
```

Please also check all other places

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 5]
```diff
@@ -987,6 +1025,32 @@ public Metadata build() {
             );
         }
 
+        private boolean validateTableOIDs(long tableOidSupplierValue) {
+            Set<Long> relationOIDs = new HashSet<>();
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Could use a `LongSet` to avoid boxing, but also makes me wonder if we need the `long` range for tables. Could go with `int` and use a BitSet here.

Cluster's probably hit scaling limits long before they'd reach the 2147483647 table+views limit

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 6]
```diff
@@ -1079,6 +1143,8 @@ public static Metadata fromXContent(XContentParser parser, boolean preserveUnkno
                         builder.version = parser.longValue();
                     } else if ("column_oid".equals(currentFieldName)) {
                         builder.columnOID = parser.longValue();
+                    } else if ("table_oid".equals(currentFieldName)) { // TODO: where is the version for writing?
+                        builder.tableOidSupplier = new DocTableInfo.OidSupplier(parser.longValue());
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: `fromXContent` is only there for BWC to read old versions where we used to store metadata in xcontent from. Newer versions use only the Writable interface. No need to add this here.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 7]
```diff
@@ -96,6 +96,7 @@
 public class Metadata implements Iterable<IndexMetadata>, Diffable<Metadata> {
 
     private static final Logger LOGGER = LogManager.getLogger(Metadata.class);
+    public static final long TABLE_OID_UNASSIGNED = 0L;
     public static final long COLUMN_OID_UNASSIGNED = 0L;
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Could have a single `OID_UNASSIGNED`

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java [hunk 8]
```diff
@@ -282,7 +282,7 @@ public ClusterState execute(ClusterState currentState) throws Exception {
 
             return indicesService.withTempIndexService(indexMetadata, indexService -> {
                 Metadata.Builder mdBuilder = Metadata.builder(currentState.metadata())
-                    .setBlobTable(relationName, indexUUID, settings, State.OPEN);
+                    .setBlobTable(relationName, indexUUID, settings, State.OPEN, currentState.metadata().tableOidSupplier().getAsLong());
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java]: Could the `Metadata.builder(currentState.metadata())` method assign the tableOidSupplier, and `setBlobTable` use it implicitly?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 9]
```diff
@@ -153,6 +154,7 @@ public interface Custom extends NamedDiffable<Custom> {
     private final long version;
     @Deprecated
     private final long columnOID;
+    private final DocTableInfo.OidSupplier tableOidSupplier;
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: `Metadata` must not have any mutable properties . `tableOidSupplier` internally has a mutable `next`.

Instead I think `Metadata` should only have a `currentMaxTableOid` and the `oidSupplier` should only be part of the builder.

File: server/src/test/java/io/crate/replication/logical/action/PublicationsStateActionTest.java [hunk 10]
```diff
@@ -410,6 +417,7 @@ public boolean hasPrivilege(Role user, Permission permission, Securable securabl
         var publication = new Publication("publisher", true, List.of());
 
         Metadata.Builder metadataBuilder = Metadata.builder();
+        metadataBuilder.tableOidSupplier(clusterService.state().metadata().tableOidSupplier());
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/replication/logical/action/PublicationsStateActionTest.java]: Given how important it is to carry over the currentOid I wonder if we should change the `builder()` method to take a parameter instead. But see also https://github.com/crate/crate/pull/18954/changes#diff-eda04a51490d14d748a884ce3577d1e7f6a0abdca8ff3207830ce74fb4345b98R419-R420

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java [hunk 11]
```diff
@@ -614,7 +614,8 @@ private void restoreRelation(RelationName relationName,
                     table.partitionedBy(),
                     table.state(),
                     Lists.concatUnique(existingTable.indexUUIDs(), indexUUIDs),
-                    table.tableVersion()
+                    table.tableVersion(),
+                    table.oid()
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: Do I get it right, that restoring existing table gets its oid as is?
I think it's allowed to restore an existing table if it's closed


Can't we have OID clash like it was with columns?

Say we have 2 clusters with this feature. Target cluster has table1 (oid = 1) and table2(oid = 2).

Source cluster has table2 (oid =1). 

So table2 on source meant to be restored on (closed) target table2, gets oid = 1 as is but it't taken by table1 on target cluster and we have a clash?

Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: Thank you for identifying an edge case, will get back to this after. BTW, are you aware of a test case for this scenario?

File: server/src/test/java/io/crate/replication/logical/action/PublicationsStateActionTest.java (hunk 10 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/replication/logical/action/PublicationsStateActionTest.java]: > But see also https://github.com/crate/crate/pull/18954/changes#diff-eda04a51490d14d748a884ce3577d1e7f6a0abdca8ff3207830ce74fb4345b98R419-R420

Thanks for reviewing. Is this link correct?

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 9 above)
Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Thank you pushed the changes, [47dfa0e](https://github.com/crate/crate/pull/18954/commits/47dfa0e513ad7b39de3d0ff58e01b77dcd2b43b1). Still need to fix tests and create a separate IntSupplier.

There are 6 call sites for `builder.tableOidSupplier()` - I think we could consider not providing `builder.tableOidSupplier()` at all?

File: server/src/main/java/org/elasticsearch/snapshots/RestoreService.java (hunk 11 above)
Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: I couldn't find a test where we restore an existing (closed) table but it's possible judging from the code
https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L484

https://github.com/crate/crate/blob/master/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L593

so what we have in docs is not really precise, needs to be fixed (needs a remark about closed tables)
https://cratedb.com/docs/crate/reference/en/latest/sql/statements/restore-snapshot.html
>If you try to restore a table that already exists, CrateDB will return an error. However, if you try to restore metadata or cluster settings that already exist, they will be overwritten.

Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: I doubled checked, on `5.10.x`, restoring snapshots to existing closed partitioned or non-partitioned tables are rejected -
```
cr> RESTORE SNAPSHOT repo.t_part3 TABLE t PARTITION (p = 'x');
OperationOnInaccessibleRelationException[The relation "doc.t" doesn't support or allow RESTORE SNAPSHOT operations, as it is currently closed.]
```

Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: But I see your concern - when restoring, we should not overwrite the current table oid with the one from snapshot.

Comment (BaurzhanSakhariev) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: >not overwrite the current table oid 

that's the question, if we can't restore to existing tables (as you checked above), there is no such thing as 
"current table", we won't event hit this code. In this case docs are good and oid usage here is good as well but it probably means that comments mentioned in https://github.com/crate/crate/pull/18954#discussion_r2735467768 are incorrect and we have some dead code?

Comment (seut) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: Seems like we artificially forbid restoring into a closed table/partition since we started to protect operations during analysis by https://github.com/crate/crate/blame/master/server/src/main/java/io/crate/metadata/table/Operation.java.
Maybe we should lift this restriction?

Comment (seut) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: In any way, I think we must assign a new table oid while restoring, the same we do for indexUUID's, and not take the OID from the snapshot as this can result in clashes when restoring a table which still exists but under a different name. When restoring a table into an existing closed one, the existing oid must be used instead of one of the snapshot, otherwise the existing table may be screwed, afaik.

Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/snapshots/RestoreService.java]: Thank you, will follow up or could be a separate feature request - it was not documented that restoring to existing closed tables is possible.

**Comment by jeeminso** [comment]
I am trying to add a test in `crate-qa` test, https://github.com/crate/crate-qa/pull/391. Please suggest if there is more testing to be done.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 12]
```diff
@@ -987,6 +1045,32 @@ public Metadata build() {
             );
         }
 
+        private boolean validateTableOIDs(int tableOidSupplierValue) {
```
Comment (jeeminso) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:1030]: We could also enable this assertions for `crate-qa` testing.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 13]
```diff
@@ -654,6 +684,12 @@ public void writeTo(StreamOutput out) throws IOException {
         }
     }
 
+    public static Builder builder(int maxTableOid) {
+        return new Builder(maxTableOid);
+    }
+
+    // TODO: will be removed, it is here so that the tests pass
+    @VisibleForTesting
     public static Builder builder() {
```
Comment (mfussenegger) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Is this planned for a follow up or can this be addressed already?

Quick way to remove might be to change the implementation to `return builder(OID_UNASSIGNED)` and then use inline refactor action. For most test that should be good enough. But some might fail with the duplicate assertion triggering.

Comment (jeeminso) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: > Quick way to remove might be to change the implementation to return builder(OID_UNASSIGNED) and then use inline refactor action.

Thank you, removed it.
